mobile/views/UserOnboarding/ExploreCommunities/index.js

```diff
@@ -0,0 +1,249 @@
+// @flow
+import * as React from 'react';
+import compose from 'recompose/compose';
+import { Easing, View, Animated } from 'react-native';
+import CommunityUpsellCards from './CommunityUpsellCards';
+import { Button } from '../../../components/Button';
+import {
+  ContinueButtonContainer,
+  ExploreCommunitiesWrapper,
+  ExploreSectionHeader,
+  ExploreSectionSubheader,
+} from './style';
+import { ViewTitle, ViewSubtitle } from '../style';
+import ExploreContext from './context';
+import ExploreCommunitiesHeader from './ExploreCommunitiesHeader';
+import ExploreSearchResults from './ExploreSearchResults';
+import addCommunityMembersMutation, {
+  type AddCommunityMembersType,
+} from '../../../../shared/graphql/mutations/communityMember/addCommunityMembers';
+
+type State = {
+  // context
+  joinedCommunities: Array<?string>,
+  onJoinCommunity: Function,
+  onLeaveCommunity: Function,
+  onSearchFocus: Function,
+  onSearchBlur: Function,
+  // local state
+  queryString: ?string,
+  isJoiningCommunities: boolean,
+};
+
+type Props = {
+  addCommunityMembers: Function,
+  refetch: Function,
+};
+
+const continueButtonHiddenY = 85;
+const continueButtonVisibleY = 0;
+
+class ExploreCommunities extends React.Component<Props, State> {
+  scrollView: any;
+
+  opacityValue = new Animated.Value(1);
+  continueButtonY = new Animated.Value(continueButtonHiddenY);
+  continueButtonOpacity = new Animated.Value(0);
+
+  constructor() {
+    super();
+    this.state = {
+      joinedCommunities: [],
+      onJoinCommunity: this.onJoinCommunity,
+      onLeaveCommunity: this.onLeaveCommunity,
+      onSearchFocus: this.onSearchFocus,
+      onSearchBlur: this.onSearchBlur,
+      onSearch: this.onSearch,
+      //
+      queryString: '',
+      isJoiningCommunities: false,
+    };
+  }
+
+  onSearchFocus = () => {
+    this.scrollView.scrollTo({ y: 0, animated: true });
+    return Animated.timing(this.opacityValue, {
+      toValue: 0.1,
+      duration: 250,
+      useNativeDriver: true,
+    }).start();
+  };
+
+  onSearchBlur = () => {
+    return Animated.timing(this.opacityValue, {
+      toValue: 1,
+      duration: 250,
+      useNativeDriver: true,
+    }).start();
+  };
+
+  onJoinCommunity = (id: string) => {
+    const { joinedCommunities } = this.state;
+    return this.setState({
+      joinedCommunities: joinedCommunities.concat(id),
+    });
+  };
+
+  onLeaveCommunity = (id: string) => {
+    const { joinedCommunities } = this.state;
+    return this.setState({
+      joinedCommunities: joinedCommunities.filter(c => c !== id),
+    });
+  };
+
+  onSearch = (queryString: string) => {
+    return this.setState({ queryString });
+  };
+
+  componentDidUpdate(_, prevState) {
```

Comment:
I think this would be an excellent use case for [`getDerivedStateFromProps`](https://reactjs.org/docs/react-component.html#static-getderivedstatefromprops), which would mean we avoid unnecessary rerenders! (updating state based on other state changes)

Reply:
How do you mean? This is derived state from state, right?

Reply:
The signature is `getDerivedStateFromProps(props, state)`, so you can derive state from state too!

Reply:
Hm, I'm not sure I want this actually - `getDerivedSTateFromProps` is meant to return an object that will apply an updated state to the component instance. In this case, however, we are only looking to dispatch actions to the animation handlers for the button. So adding a new state field for `continueButtonVisible` is duplicitous.

Reply:
Oh I see, sorry—I thought `showContinueButton` was setting state! Nevermind then, this is good 👌